### ISSUE TITLE
io-wait used 100% cpu after 6.1.39 and 6.5.1

### ISSUE THREAD

**Issue by beldzhang** [axboe/liburing#943]
after this commit:
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=v6.1.39&id=f32dfc802e8733028088edf54499d5669cb0ef69
a running io-uring object will cause one cpu showing 100% usage of io-wait, in my environment 8 rings/threads make 8 cpu got 100% io-wait usage. beside this nothing else found.

try to revert this commit from 6.1.39 then everything is ok.
6.1.51/6.5.1 it is the same, can not direct revert this commit so not tested yet.

all following test got the same performance result:
6.1.38
6.1.39
6.1.39(reverted)
6.1.51

re-produce:
use this testing program: https://github.com/axboe/liburing/files/9571382/issue-643-v2.zip, build and just run server, it will showing io-wait 100% on one cpu.

**Comment by redbaron** [comment]
isn't IO wait time is what you want to see? That is time app is blocked waiting on io (at least it was the case for file IO), but CPU has nothing else to do, so that time is accounted as IO wait.

**Comment by isilence** [comment]
Note that io-wait doesn't burn CPU cycles, it's sleeping and so it's not a problem apart from reporting. There was a change doing that, and I think it actually makes more sense reporting a task waiting for io_uring completions as io-wait

**Comment by axboe** [comment]
Yes this is expected. iowait literally just means "waiting on IO", which is what the task is doing. It does NOT mean that it's busy 100% of the time, in fact if you have 100% iowait it means 1 task is sleeping waiting on IO 100% of the time.

**Comment by beldzhang** [comment]
> Yes this is expected. iowait literally just means "waiting on IO", which is what the task is doing. It does NOT mean that it's busy 100% of the time, in fact if you have 100% iowait it means 1 task is sleeping waiting on IO 100% of the time.

noticed that, when this happen, system load is only 0.0x

we are using io-uring in a storage service, so is very sensitive about the storage/network load, previously io-wait is a good indicator to check this.
and before kernel 5.0, iostat's %util value can be used also, but after 5.0, this number will easily up to 100% even on a small load.
is there any other way to check the disk load?

and I calculate a performance score after each test based on total cpu usage, just ignore the io-wait part looks like not a good solution...

**Comment by beldzhang** [comment]
looks no more comments, closed.

**Comment by romange** [comment]
@axboe, one of Dragonfly's users, also reported this as a behavioural change: Dragonfly, which does not use disk IO, bumps up the IOWAIT metric to 100%. If they run it using the epoll API, it does not affect IOWAIT. I am just confirming whether this change is indeed intended.

**Comment by rickytato** [comment]
I see same issue with Kernel 6.5.11 (on Proxmox)
![image](https://github.com/axboe/liburing/assets/353896/d357af70-fb8b-4d13-9496-9ee665d3a13c)

**Comment by beldzhang** [comment]
emmm... reopen?

**Comment by RX14** [comment]
IOwait has been traditionally thought of as "waiting for disk io", which will always complete. Since io_uring can be used to wait on the network, which has unbounded waiting time, it changes the metric considerably. For example, many monitoring systems have alerts for iowait being high, correctly or not assuming it to be a proxy for disk contention.

**Comment by axboe** [comment]
Here's what I think we should do:

1) Default to not using iowait, as it is indeed somewhat confusing for networked or mixed network/storage workloads. I do think iowait is an awful metric that makes very little sense for async workloads, even just pure storage based ones. Lots of consumers will assume it's busy time, or has a direct correlation with disk usage, which is just wrong.
2) Add an IORING_ENTER_IOWAIT flag that can be used in conjunction with IORING_ENTER_GETEVENTS. If set, iowait will be used. Storage can use this, if they so wish.
3) Add an IORING_FEAT_IOWAIT flag, which tells the app/liburing that this feature is available.
4) Add liburing helpers, ala io_uring_set_iowait() and io_uring_clear_iowait(), which can be used to toggle this flag, if IORING_FEAT_IOWAIT is set. Storage based workloads can set this.

And that should be it. That gives the app control over whether iowait should be used or not.

**Comment by beldzhang** [comment]
will test when avaliable

**Comment by isilence** [comment]
@axboe, I just mentioned it in the mailing list, but even though I don't understand why people are taken aback by hi iowait from io_uring waiting, but I think we should just revert that change, there has been too many reports from different people regarding this one. We should be able to do the optimisation that was the reason for the change without reporting iowait.

**Comment by axboe** [comment]
We can't just revert it, as it solved a real problem. I have my doubts that we can separate the cpufreq side from iowait in a way that would make the scheduler side happy. If we can, I'd be all for it, and would love to see a patch.

**Comment by romange** [comment]
A great discussion about this topic on [lore.kernel.org](https://lore.kernel.org/all/05dc2185-2935-4eac-a8e7-f407035f9315@kernel.dk/T/#mc2fa25c51f5ff42a8b35de95a8ce6f3c40b23d33).
 
Just for us to understand, once we call `io_uring_register_iowait`, it will flag networking I/O as iowait  but iouring will run in more efficient manner?

Another interesting comment I read is about multiple rings. Currently https://github.com/romange/helio has ring-per-thread architecture. @axboe  are you saying that sometimes it makes sense to have two rings? For what use-cases it makes sense?

**Comment by isilence** [comment]
> A great discussion about this topic on [lore.kernel.org](https://lore.kernel.org/all/05dc2185-2935-4eac-a8e7-f407035f9315@kernel.dk/T/#mc2fa25c51f5ff42a8b35de95a8ce6f3c40b23d33).
> 
> Just for us to understand, once we call `io_uring_register_iowait`, it will flag networking I/O as iowait but iouring will run in more efficient manner?

The long story. There is a patch upstream since a while ago which does two unrelated things: first it enables some cpu governor optimisation useful for QD1 and not only, and it also changes the io-wait stat behaviour as per this thread. They're coupled together for implementation reasons, it's much easier going this way. So, the optimisation is already in the kernel and always enabled, let's say it's a free lunch. Now, that `io_uring_register_iowait()` patch would disable the optimisation by default and turn it back on only if you call the function.

I have to say that it's quite a horrendous approach, having side effects from seemingly an optimisation, mixing responsibilities and levels at what the feature enabled and the iowait stat is observed, and so on. I think the register_iowait patch should never be given the light, at least as far as it mixes things together.

**Comment by isilence** [comment]
> Another interesting comment I read is about multiple rings. Currently https://github.com/romange/helio has ring-per-thread architecture. @axboe are you saying that sometimes it makes sense to have two rings? For what use-cases it makes sense?

IMHO, it doesn't make sense apart maybe from some IOPOLL + normal ring weird cases. However, sometimes it happens  (unfortunately). For instance, when a library / framework you use has some io_uring support inside, and then the app creates another rings for its own purposes.

**Comment by axboe** [comment]
There will be no register iowait, the current pending fixes are here:

https://git.kernel.dk/cgit/linux/log/?h=iowait.2

and will be posted for review soon, so they can get into the 6.10 kernel.

**Comment by beldzhang** [comment]
> https://git.kernel.dk/cgit/linux/log/?h=iowait.2

@axboe 
brief tested, io-wait is gone, will following up. mailing list followed also.

@isilence
to the end users, they are sensitive to the latency and high response time of server, to sys admin, io-wait and load are directly showing the situations. generally storage parts is slowest in whole system, user/admin didn't care about the waiting of io read/write is sync or async, they just want to know how much loading of the *entire server*
I already remove iostat %util displaying because it's non-sense anymore. but 100% iowait of io-uring on cpu, is terrify a *lot* of users/admins.

**Comment by beldzhang** [comment]
ready for testing,  for-6.10/io_uring? for-6.10/block or for-next? thanks.

**Comment by beldzhang** [comment]
emmm.... any updates?

**Comment by solarvm** [comment]
still happening for us too

**Comment by isilence** [comment]
Nothing was merged yet, as it's a low priority reporting issue. However, there is interest in that for some other reasons, and it's in the backlog. will get picked up hopefully soon.

**Comment by beldzhang** [comment]
`iowait.4` branch tested, no iowait usage. detailed test pending, no regression for now.

**Comment by jurim76** [comment]
So what's next?
Wast this branch merged into master? Is this fix available in a new liburing version?

**Comment by Athinivas** [comment]
@beldzhang  when will this fix be merged into master?.. could be very helpful if you provide a estimated time for same... thanks

**Comment by beldzhang** [comment]
emmm.... it's not by me.. I'm waiting also... latest message was Set 2024, some patches in mailing-list but I have no idea how to test :(

**Comment by beldzhang** [comment]
https://web.git.kernel.org/pub/scm/linux/kernel/git/axboe/linux-block.git/commit/?h=for-next&id=1c776ea242587a50e93a5821c42a64243f6bbdca

tested for-next and backport to 6.12.19, iowait is gone
functionality testing only, performance test pending.

I preferred this way instead of previous API one.

**Comment by axboe** [comment]
I've ranted about this before, but the problem here is really ignorant monitoring software or system dashboards that equate iowait with busy time. It's no such thing, has never been, nor does it tell you ANYTHING about disk load on the system. Or anything else useful, for that matter.

**Comment by axboe** [comment]
@beldzhang Also curious on your performance testing, do report!

**Comment by beldzhang** [comment]
6.12.16： vanilla kernel; 6.12.19: backported 1c776ea2
10 Gbps ethernet, 4k random read from nvme raw dev, connection: 1000, 16 connection per ring
whole system statistics for 3 minitus
![Image](https://github.com/user-attachments/assets/96a19348-6e77-4011-aa59-516664225ebd)

a brief comparison, looks like disable iowait got better performance, based on cpu:total and net:rx/tx.

**Comment by axboe** [comment]
How stable are these numbers?

**Comment by isilence** [comment]
What is the meaning behind "cpu:total"?

**Comment by beldzhang** [comment]
> How stable are these numbers? 
> Edit: I mean, if you reboot and run any of these again, what's the expected deviation for them?

oh, I forgot can do this test just turn on iowait in /proc/../io_uring_iowait
each statistics was collected after buffer/cache got full, that means after running about 1~2 min then start collection
my server has 128G ram, locked 96G, then 32G left for block buffer

> What is the meaning behind "cpu:total"?

ctxt, intr, cpu - read from /proc/stat, calc percentage from begin to end.
net rx/tx - read from /proc/net/dev, calc avg speed (MB/s) over 10Gbps link.
proc - read from /proc/<server>/stat, calc percentage, split to user and sys part

**Comment by isilence** [comment]
> > What is the meaning behind "cpu:total"?
> 
> ctxt, intr, cpu - read from /proc/stat, calc percentage from begin to end. net rx/tx - read from /proc/net/dev, calc avg speed (MB/s) over 10Gbps link. proc - read from /proc//stat, calc percentage, split to user and sys part

Let me elaborate, you said it's better based on "cpu:cycles" and tx/rx, but what does "cpu:cycles" actually signify here? For me it looks like a nonsense metric.

**Comment by axboe** [comment]
> Let me elaborate, you said it's better based on "cpu:cycles" and tx/rx, but what does "cpu:cycles" actually signify here? For me it looks like a nonsense metric.

It looks like it's attempting to calculate the CPU busy percentage - but it includes iowait, which of course isn't a busy metric at all and should never (ever) be used for something like that. If you disregard iowait from that line, the cpu numbers come out identical (+/- 0.01).

**Comment by beldzhang** [comment]
about our server
- it's a iscsi-like storage protocol, tcp with long-time connection, most time with low rate(even to zero), burst rate very high(10Gbps or 25Gbps)
- concurrent designed to 5000+, planned io-uring replace old thread-model

about iowait
  - for me iowait is ok, there is no need to fix, for our customers, it doesn't
  
  - about 20+ years ago, servers are with rotational disks, disk delays cause high iowait and high load values, server will response very slow and client looks hangup.
  so the users are very sensitive with the iowait value. that make our transfer to io-uring model 2+ years delay because user are scaried when they saw big iowait.
  for now, most server are with solid disks, which is much fast than nic speed, but user's mind can not keepup so quickly...

  - we could remove iowait monitoring in our software, sadly there are a lot of tools still showing, ex: top, iostat
  and as others mentioned, some db and load-balance software depends on iowait.
  
  - if hard to resolv, I could keep this patch as private just for our kernel builds to make users feel better :p.

about our testing-tool
  - it is not very reliable, mostly just for compare side-by-side for changes in kernel and server
  - check factors are: 1) higher net rate;  2) higher disk read/write;  3) lower cpu and load.
  - power consumer not considered as our case.

**Comment by isilence** [comment]
> > Let me elaborate, you said it's better based on "cpu:cycles" and tx/rx, but what does "cpu:cycles" actually signify here? For me it looks like a nonsense metric.
> 
> It looks like it's attempting to calculate the CPU busy percentage - but it includes iowait, which of course isn't a busy metric at all and should never (ever) be used for something like that. If you disregard iowait from that line, the cpu numbers come out identical (+/- 0.01).

Right, and that's what I'm trying to say. iowait is by definition a category of idle, it doesn't burn the cpu. I can't even think of an example how "cpu:cycles" calculated in this way can be used or for what purpose. Not trying to protect the iowait boost, it can affect many metrics in different ways, it can even increase power consumption with the same CPU util, but idle <-> iowait __stat__ changes are purely cosmetic.

**Comment by isilence** [comment]
> about iowait
> 
>     * for me iowait is ok, there is no need to fix, for our customers, it doesn't
> 
>     * about 20+ years ago, servers are with rotational disks, disk delays cause high iowait and high load values, server will response very slow and client looks hangup.
>       so the users are very sensitive with the iowait value. that make our transfer to io-uring model 2+ years delay because user are scaried when they saw big iowait.
>       for now, most server are with solid disks, which is much fast than nic speed, but user's mind can not keepup so quickly...
> 
>     * we could remove iowait monitoring in our software, sadly there are a lot of tools still showing, ex: top, iostat
>       and as others mentioned, some db and load-balance software depends on iowait.
> 
>     * if hard to resolv, I could keep this patch as private just for our kernel builds to make users feel better :p.

I hear you, and the patch will likely get merged soon unless we break the gatekeeping wall on the kernel sched/cpufreq side and do it a better way. However, I hope to stop people from misinterpreting the iowat stat here, as I've seen many having pretty wild ideas about it.

**Comment by axboe** [comment]
> For me, CPU always sits at the lowest idle frequency, despite having 100% iowait.

Sounds like a machine/setup issue. It should avoid the lower power states on the CPU where that task is currently waiting. This isn't related to io_uring at all, io_uring just marks it as being in iowait, just like you'd see if iowait was used elsewhere.

Eg if you do:

```
# taskset -c 0 dd if=/dev/nvmeX of=/dev/null bs=4k iflag=direct
```

and watch the C state of CPU 0, then it should not be in lower power states.

**Comment by axboe** [comment]
Agree, sounds like it's working as expected then, for longer waits you very much do want to go lower in power. This should be the same then for anything waiting off io_uring_enter(2) with GETEVENTS set, that should also drift down in power states depending on the duration.

**Comment by ValeZAA** [comment]
>But if I change it to say bs=4M, I have 100% iowait with CPU remaining at the lowest state.

So there is no bug and this can be closed? What the patches for linux kernel for? It will be still be the same, just hidden?

**Comment by HanabishiRecca** [comment]
> So there is no bug and this can be closed?

Well, `dd` performs actual work here, it reads from the disk at max speed. So iowait metric is actually useful, shows how busy the disk I/O is.

But io_uring shows 100% iowait when literally nothing is happening. E.g. a server app simply waiting for incoming network requests.

> What the patches for linux kernel for? It will be still be the same, just hidden?

Yes.

**Comment by HanabishiRecca** [comment]
> patch already in kernel 6.15-rc1, [07754bfd9aee59063f8549f6e4d455eae636ecc7](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?h=v6.15-rc1&id=07754bfd9aee59063f8549f6e4d455eae636ecc7)

No global sysfs knob after all, huh. That's unfortunate.

**Comment by Athinivas** [comment]
@beldzhang  Has the patch been released? anything decided ? Thanks!

**Comment by axboe** [comment]
Closing.
